### PR TITLE
Mwolters/1611

### PR DESCRIPTION
**Pull request by MarkWolters**
buffer metrics that go out of scope before end of test at the session level and print to screen at end of session.

See https://github.com/nosqlbench/nosqlbench/issues/1611

### CODE REVIEW

File: nb-api/src/main/java/io/nosqlbench/components/NBBaseComponent.java [hunk 1]
```diff
@@ -122,6 +126,14 @@ public final void close() throws RuntimeException {
         }
     }
 
+    public void reportExecutionMetric(NBMetric m) {
+        if (parent != null) {
+            parent.reportExecutionMetric(m);
+        } else {
+            logger.debug("buffering metric " + m.toString() + " for final report");
```
Comment (jshook) [inline nb-api/src/main/java/io/nosqlbench/components/NBBaseComponent.java:133]: This log line doesn't match with any other logic I can see here.

Comment (MarkWolters) [reply nb-api/src/main/java/io/nosqlbench/components/NBBaseComponent.java]: good point. Changed it to a warn level message about encountering a null parent outside of NBSession.

File: nb-api/src/main/java/io/nosqlbench/components/NBBaseComponent.java [hunk 2]
```diff
@@ -122,6 +126,14 @@ public final void close() throws RuntimeException {
...
     }
 
+    public void reportExecutionMetric(NBMetric m) {
+        if (parent != null) {
+            parent.reportExecutionMetric(m);
+        } else {
+            logger.warn(() -> "Component " + this.toString() + " has a non-session null parent component");
+        }
+    }
+
```
Comment (ShaunakDas88) [inline nb-api/src/main/java/io/nosqlbench/components/NBBaseComponent.java:136]: This is perhaps a dumb question, apologies in advance. 

Each `NBComponent` has a single parent `NBComponent`, and the root of the tree corresponds to the Session, correct? 

If so, this is baiscally just parenting up to the root/Session, and once we reach the root/Session, we just emit the warning message?

Comment (ShaunakDas88) [reply nb-api/src/main/java/io/nosqlbench/components/NBBaseComponent.java:136]: Never mind, just realized we override this method in `NBSession`.

**Comment by ShaunakDas88** [comment]
Overall makes sense. One thing I wanted to raise for discussion (and perhaps this is a bigger discussion for elsewhere, that's fine). 

Would it make sense to also store reference to appropriate root/`NBSession` for each `NBComponent`? We currently keep references to parents. I don't know how deep we expect the tree to get, but having such would allow for skipping directly up to the `NBSession`.

**Comment by MarkWolters** [comment]
@jshook it was just waiting for merge, I'll go ahead and merge it now.

**Pull request merged**